### PR TITLE
Add ignoreProperties option to unit-blacklist and unit-whitelist

### LINKED ISSUE #1062: Add ignoreProperties: {} option to unit-blacklist and unit-whitelist
If you have a global rule to say blacklist px there is no current way to add an exception ie blacklist px globally except on borders, text-shadow, etc. as by whitelisting these in property-unit-whitelist the rule is overwritten. I assume it is the same with  unit-whitelist and property-unit-blacklist (although have not used it this way around so cannot verify).

I would either expect you can add exceptions to the global blacklist or the property-unit acts like an exception.

> Which rule, if any, is this issue related to?

'unit-blacklist'
'property-unit-whitelist'

'unit-whitelist'
'property-unit-blacklist'

> What CSS is needed to reproduce this issue?
 Both of these will error even though we are asking for the px on border
e.g.

```css
.foo {
  border: 1px solid purple;
}

.bar {
  border: 1rem solid purple;
}
```

> What stylelint configuration is needed to reproduce this issue?

e.g.

```json
    "unit-blacklist": ["px"],

    "property-unit-blacklist": {
      "/^border/": ["em", "rem", "%"]
    },
    "property-unit-whitelist": {
      "/^border/": ["px"]
    },
```

> Which version of stylelint are you using?

'5.4.0'

> How are you running stylelint: CLI, PostCSS plugin, Node API?

SublimeText 3

> What did you expect to happen?

No warnings when border uses px

> What actually happened (e.g. what warnings or errors you are getting)?

The unit is blacklisted

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/32903924-add-ignoreproperties-option-to-unit-blacklist-and-unit-whitelist?utm_campaign=plugin&utm_content=tracker%2F9282518&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F9282518&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by gaidarenko**
Closes #1062

### CODE REVIEW

File: src/rules/unit-blacklist/index.js [hunk 1]
```diff
@@ -52,6 +64,26 @@ function rule(blacklistInput) {
...
+/* Accept only objects like this:
+ *
+ * ignoreProperties = {
+ *   value1: ["item11", "item12", "item13"],
+ *   value2: ["item21", "item22", "item23"],
+ *   value3: ["item31", "item32", "item33"],
+ * }
+ *
+ */
+function validateIgnorePropertiesOption(options) {
```
Comment (evilebottnawi) [inline src/rules/unit-blacklist/index.js]: Be good create `utils` for this, because we have repeated code in two rules

**Comment by ntwb** [comment]
@gaidarenko Thanks, we also need an entry in the changelog noting this new rule option

http://stylelint.io/developer-guide/rules/#adding-options-to-existing-rules

**Comment by gaidarenko** [comment]
I moved repeating code into `utils` and added changes to `CHANGELOG`.

File: src/rules/unit-blacklist/README.md [hunk 2]
```diff
@@ -49,3 +49,48 @@ a { height: 100vmin; }
 ```css
 a { animation: animation-name 5s ease; }
 ```
+
+## Optional Options
```
Comment (davidtheclark) [inline src/rules/unit-blacklist/README.md]: Please double-check against the current documetnation that this is the title we're using.

File: src/rules/unit-blacklist/README.md [hunk 3]
```diff
@@ -49,3 +49,48 @@ a { height: 100vmin; }
...
+
+### `ignoreProperties: { unit: ["property", "/regex/"] }`
+
+Ignore the specified unites in the specified properties.
```
Comment (davidtheclark) [inline src/rules/unit-blacklist/README.md]: units

Comment (davidtheclark) [inline src/rules/unit-blacklist/README.md]: How about this: `Ignore units in the values of declarations with the specified properties.`

File: src/rules/unit-blacklist/README.md [hunk 4]
```diff
@@ -49,3 +49,48 @@ a { height: 100vmin; }
...
+Given:
+
+```js
+["px", "vmin"]
+
+ignoreProperties: {
+  "px": [ "font-size", "/^border/" ],
+  "vmin": [ "width" ]  
+}
+```
```
Comment (davidtheclark) [inline src/rules/unit-blacklist/README.md]: This is unclear. Can you please check for precedent in existing documentation and match that format?

File: src/rules/unit-blacklist/README.md [hunk 5]
```diff
@@ -49,3 +49,48 @@ a { height: 100vmin; }
...
+
+```css
+a { width: 100vmin; }
+```
+
+The following patterns are considered warnings:
+
+```css
+a { line-height: .1px; }
```
Comment (davidtheclark) [inline src/rules/unit-blacklist/README.md]: Please remove the decimal point. It's confusing.

File: src/rules/unit-blacklist/README.md [hunk 6]
```diff
@@ -49,3 +49,48 @@ a { height: 100vmin; }
...
+```
+
+```css
+a { -moz-border-radius-topright: 40px; }
```
Comment (davidtheclark) [inline src/rules/unit-blacklist/README.md]: Should there be a hyphen between `top-right`?

File: src/rules/unit-blacklist/__tests__/index.js [hunk 7]
```diff
@@ -196,3 +196,48 @@ testRule(rule, {
...
 })
+
+testRule(rule, {
+  ruleName,
+
+  config: [ [
+    "px",
+    "vmin",
+  ], { ignoreProperties: { "px": [ "font-size", "margin", "/^border/" ],
+                           "vmin": [ "width", "height" ] } } ],
```
Comment (davidtheclark) [inline src/rules/unit-blacklist/__tests__/index.js]: Please match the formatting for configs in other tests.

File: src/rules/unit-blacklist/index.js [hunk 8]
```diff
@@ -33,6 +41,11 @@ function rule(blacklistInput) {
 
         if (!unit || (unit && blacklist.indexOf(unit.toLowerCase()) === -1)) { return }
 
+        if (options && node.prop) {
+          const props = _.get(options, `ignoreProperties.${unit.toLowerCase()}`, [])
+          if (matchesStringOrRegExp(node.prop, props) !== false) { return }
```
Comment (davidtheclark) [inline src/rules/unit-blacklist/index.js]: Please use the `optionsMatches` util.

File: src/rules/unit-whitelist/README.md [hunk 9]
```diff
@@ -53,3 +53,48 @@ a { height: 100PX; }
...
+a { margin: 0 20rem; }
+```
+
+```css
+a { -moz-border-radius-topright: 20rem; }
+```
+
+```css
+a { height: 100%; }
+```
```
Comment (davidtheclark) [inline src/rules/unit-whitelist/README.md]: Adjust this doc in the same way you adjust the above.

File: src/rules/unit-whitelist/__tests__/index.js [hunk 10]
```diff
@@ -194,3 +194,48 @@ testRule(rule, {
...
 })
+
+testRule(rule, {
+  ruleName,
+
+  config: [ [
+    "px",
+    "em",
+  ], { ignoreProperties: { "rem": [ "line-height", "margin", "/^border/" ],
+                           "%": [ "width", "height" ] } } ],
```
Comment (davidtheclark) [inline src/rules/unit-whitelist/__tests__/index.js]: Needs reformatting.

File: src/rules/unit-whitelist/index.js [hunk 11]
```diff
@@ -33,6 +41,11 @@ function rule(whitelistInput) {
 
         if (!unit || (unit && whitelist.indexOf(unit.toLowerCase()) !== -1)) { return }
 
+        if (options && node.prop) {
+          const props = _.get(options, `ignoreProperties.${unit.toLowerCase()}`, [])
+          if (matchesStringOrRegExp(node.prop, props) !== false) { return }
```
Comment (davidtheclark) [inline src/rules/unit-whitelist/index.js]: Same as above: use the `optionsMatches` util.

**Comment by davidtheclark** [comment]
@gaidarenko Great job making the util. I left some other code comments for you.

**Comment by gaidarenko** [comment]
@davidtheclark I adjusted the code according to your comments.

File: src/rules/unit-blacklist/__tests__/index.js [hunk 12]
```diff
@@ -196,3 +196,45 @@ testRule(rule, {
     column: 18,
   }],
 })
+
+testRule(rule, {
+  ruleName,
+
+  config: [ [ "px", "vmin" ], { ignoreProperties: { "px": [ "font-size", "margin", "/^border/" ],
+                                                  "vmin": [ "width", "height" ] } } ],
```
Comment (davidtheclark) [inline src/rules/unit-blacklist/__tests__/index.js]: Please look at existing tests and reformat accordingly.

**Comment by davidtheclark** [comment]
@gaidarenko Great. I still want to see the config in tests formatted more like what we do elsewhere. Otherwise all seems good to me.

**Comment by davidtheclark** [comment]
@jeddy3 I can just adjust the nitpicky formatting and merge unless you have any reservations about this branch.

**Comment by davidtheclark** [comment]
Merged as e0b1d22daa08fcdfd4a722a73a5f7ccc5cd1de16 and e0b1d22daa08fcdfd4a722a73a5f7ccc5cd1de16. Thanks @gaidarenko!

**Pull request closed without merging**